### PR TITLE
Don't bind named members on missing attribute

### PR DESCRIPTION
**Pull request by jcouv**
Small fix to remove some unhelpful cascading diagnostics.
Fixes https://github.com/dotnet/roslyn/issues/68370

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs [hunk 1]
```diff
@@ -503,11 +503,21 @@ private BoundAssignmentOperator BindNamedAttributeArgument(AttributeArgumentSynt
         {
             bool wasError;
             LookupResultKind resultKind;
-            Symbol namedArgumentNameSymbol = BindNamedAttributeArgumentName(namedArgument, attributeType, diagnostics, out wasError, out resultKind);
+            Symbol? namedArgumentNameSymbol;
 
-            ReportDiagnosticsIfObsolete(diagnostics, namedArgumentNameSymbol, namedArgument, hasBaseReceiver: false);
+            if (attributeType.IsErrorType())
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:508]: Should we just return BadExpression from this code path?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:508]: Oh, I see what you're saying. We can still bind the RHS value and potentially get a simpler quick exit. Let me try.

**Review by RikkiGibson — APPROVED** [review]

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

File: src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs [hunk 2]
```diff
@@ -501,10 +501,19 @@ protected bool IsAttributeConditionallyOmitted(NamedTypeSymbol attributeType, Sy
 
         private BoundAssignmentOperator BindNamedAttributeArgument(AttributeArgumentSyntax namedArgument, NamedTypeSymbol attributeType, BindingDiagnosticBag diagnostics)
         {
+            Debug.Assert(namedArgument.NameEquals is not null);
+            IdentifierNameSyntax nameSyntax = namedArgument.NameEquals.Name;
+
+            if (attributeType.IsErrorType())
+            {
+                var badLHS = BadExpression(nameSyntax, lookupResultKind: LookupResultKind.Empty);
+                var badRHS = this.BindValue(namedArgument.Expression, diagnostics, BindValueKind.RValue);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:510]: This needs a `BindToNaturalTypeForErrorRecovery`. Please add a test with a target-typed node such as a switch expression, and pull on the value through some path that will fail if it's not bound to a real type, such as IOperation.

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 2)

File: src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs]: Good catch. I thought the `null` already covered the scenario for typeless expressions, but the switch indeed asserted. Thanks

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests.cs [hunk 3]
```diff
@@ -11291,6 +11291,58 @@ class A<T> : System.Attribute
...
+                // (1,22): error CS0246: The type or namespace name 'InexistentAttribute' could not be found (are you missing a using directive or an assembly reference?)
+                // [assembly: /*<bind>*/Inexistent(SomeProperty = 1, F = null, G = 0 switch { _ => 1 })/*</bind>*/]
+                Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "Inexistent").WithArguments("InexistentAttribute").WithLocation(1, 22),
+                // (1,22): error CS0246: The type or namespace name 'Inexistent' could not be found (are you missing a using directive or an assembly reference?)
+                // [assembly: /*<bind>*/Inexistent(SomeProperty = 1, F = null, G = 0 switch { _ => 1 })/*</bind>*/]
+                Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "Inexistent").WithArguments("Inexistent").WithLocation(1, 22)
+            };
+
+            VerifyOperationTreeAndDiagnosticsForTest<AttributeSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests.cs:11345]: >} [](http://example.com/codeflow?start=8&length=1)

Consider also adding explicit semantic model verification:
```
            var comp = CreateCompilation(source);
            var tree = comp.SyntaxTrees[0];
            var model = comp.GetSemanticModel(tree);
            var one = tree.GetRoot().DescendantNodes().OfType<LiteralExpressionSyntax>().First();

            Assert.Equal("1", one.ToString());
            Assert.Equal("SomeProperty = 1", one.Parent.ToString());
            Assert.Equal("System.Int32", model.GetTypeInfo(one).Type.ToTestDisplayString());
```

Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests.cs:11345]: I don't think the semantic model validation adds much on top of IOperation check, which already shows how everything got bound.

**Pull request merged**